go-controller/pkg/ovn/libovsdbops/portgroup.go

```diff
@@ -0,0 +1,344 @@
+	}
+
+	return err
+}
+
+func BuildPortGroup(hashName, name string, ports []*nbdb.LogicalSwitchPort, acls []*nbdb.ACL) *nbdb.PortGroup {
+	var aclUUIDs []string
+	if len(acls) > 0 {
+		aclUUIDs = make([]string, 0, len(acls))
+		for _, acl := range acls {
+			aclUUIDs = append(aclUUIDs, acl.UUID)
+		}
+	}
+
+	var portUUIDs []string
+	if len(ports) > 0 {
+		portUUIDs = make([]string, 0, len(ports))
+		for _, port := range ports {
+			portUUIDs = append(portUUIDs, port.UUID)
+		}
+	}
+
+	return &nbdb.PortGroup{
+		Name:        hashName,
+		ACLs:        aclUUIDs,
+		Ports:       portUUIDs,
+		ExternalIDs: map[string]string{"name": name},
+	}
+}
+
+func createOrUpdatePortGroupOps(nbClient libovsdbclient.Client, ops []libovsdb.Operation, pg *nbdb.PortGroup) ([]libovsdb.Operation, error) {
+	if ops == nil {
+		ops = []libovsdb.Operation{}
+	}
+
+	err := findPortGroup(nbClient, pg)
+	if err != nil && err != libovsdbclient.ErrNotFound {
+		return nil, err
+	}
+
+	if err == libovsdbclient.ErrNotFound {
+		op, err := nbClient.Create(pg)
+		if err != nil {
+			return nil, err
+		}
+		ops = append(ops, op...)
+		return ops, nil
+	}
+
+	mutations := []model.Mutation{}
+	if len(pg.Ports) > 0 {
+		mutations = append(mutations, model.Mutation{
+			Field:   &pg.Ports,
+			Mutator: libovsdb.MutateOperationInsert,
+			Value:   pg.Ports,
+		})
+	}
+	if len(pg.ACLs) > 0 {
+		mutations = append(mutations, model.Mutation{
+			Field:   &pg.ACLs,
+			Mutator: libovsdb.MutateOperationInsert,
+			Value:   pg.ACLs,
+		})
+	}
+	if len(pg.ExternalIDs) > 0 {
+		mutations = append(mutations, model.Mutation{
+			Field:   &pg.ExternalIDs,
+			Mutator: libovsdb.MutateOperationInsert,
+			Value:   pg.ExternalIDs,
+		})
+	}
+	if len(mutations) > 0 {
+		op, err := nbClient.Where(pg).Mutate(pg, mutations...)
+		if err != nil {
+			return nil, err
+		}
+		ops = append(ops, op...)
+		return ops, nil
+	}
+
+	return ops, nil
+}
+
+func CreateOrUpdatePortGroupsOps(nbClient libovsdbclient.Client, ops []libovsdb.Operation, pgs ...*nbdb.PortGroup) ([]libovsdb.Operation, error) {
+	if ops == nil {
+		ops = []libovsdb.Operation{}
+	}
+	for _, pg := range pgs {
+		var err error
+		ops, err = createOrUpdatePortGroupOps(nbClient, ops, pg)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return ops, nil
+}
+
+func CreateOrUpdatePortGroups(nbClient libovsdbclient.Client, pgs ...*nbdb.PortGroup) error {
+	ops, err := CreateOrUpdatePortGroupsOps(nbClient, nil, pgs...)
+	if err != nil {
+		return err
+	}
+
+	_, err = TransactAndCheck(nbClient, ops)
+	return err
+}
+
+func AddPortsToPortGroupOps(nbClient libovsdbclient.Client, ops []libovsdb.Operation, name string, ports ...string) ([]libovsdb.Operation, error) {
+	if ops == nil {
+		ops = []libovsdb.Operation{}
+	}
+
+	pg := &nbdb.PortGroup{
+		Name: name,
+	}
+	err := findPortGroup(nbClient, pg)
+	if err != nil && err != libovsdbclient.ErrNotFound {
+		return nil, err
+	}
+
+	// ignore possible ErrNotFound, attempt by name anyway
```

Comment:
We attempt the Mutate using the name as index instead of the UUID. The cache is asynchronous so it is possible that we don't have the PG in cache but the operation might still succeed server side.

Reply:
Hmm, didn't that change once they fixed: https://github.com/ovn-org/libovsdb/issues/212 ? If anything has already created the port group and called `Transact` on that operation, then we should be able to be certain that the cache is updated with the item once `Transact` has returned, right? Hence, at this point if `findPortGroup` returns `libovsdbclient.ErrNotFound` we should create the port group, not skip the error - because it's really not existent.